haystack_experimental/components/converters/image_utils.py

```diff
@@ -0,0 +1,301 @@
+
+import base64
+from io import BytesIO
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from haystack import logging
+from haystack.dataclasses import ByteStream
+from haystack.lazy_imports import LazyImport
+
+with LazyImport("Run 'pip install pypdf pdf2image'") as pypdf_and_pdf2image_import:
+    import pdf2image
+    from pypdf import PdfReader
+
+with LazyImport("Run 'pip install pillow'") as pillow_import:
+    from PIL import Image as PILImage
+    from PIL.Image import Image
+    from PIL.ImageFile import ImageFile
+
+
+logger = logging.getLogger(__name__)
+
+
+_OPENAI_DETAIL_TO_IMAGE_SIZE = {"low": (512, 512), "high": (768, 2_048), "auto": (768, 2_048)}
+
+# NOTE: We have to rely on this since our util functions are using the bytestream object.
+#      We could change this to use the file path instead, where the file extension is used to determine the format.
+# This is a mapping of image formats to their MIME types.
+# from PIL import Image
+# Image.init()  # <- Must force all plugins to initialize to get this mapping
+# print(Image.MIME)
+FORMAT_TO_MIME = {
+    "BMP": "image/bmp",
+    "DIB": "image/bmp",
+    "PCX": "image/x-pcx",
+    "EPS": "application/postscript",
+    "GIF": "image/gif",
+    "PNG": "image/png",
+    "JPEG2000": "image/jp2",
+    "ICNS": "image/icns",
+    "ICO": "image/x-icon",
+    "JPEG": "image/jpeg",
+    "MPEG": "video/mpeg",
+    "TIFF": "image/tiff",
+    "MPO": "image/mpo",
+    "PALM": "image/palm",
+    "PDF": "application/pdf",
+    "PPM": "image/x-portable-anymap",
+    "PSD": "image/vnd.adobe.photoshop",
+    "SGI": "image/sgi",
+    "TGA": "image/x-tga",
+    "WEBP": "image/webp",
+    "XBM": "image/xbm",
+    "XPM": "image/xpm",
+}
+MIME_TO_FORMAT = {v: k for k, v in FORMAT_TO_MIME.items()}
+# Adding some common MIME types that are not in the PIL mapping
+MIME_TO_FORMAT["image/jpg"] = "JPEG"
+
+
+def encode_image_to_base64(
+    bytestream: ByteStream,
+    size: Optional[Tuple[int, int]] = None,
+) -> Tuple[str, str]:
+    """
+    Encode an image from a ByteStream into a base64-encoded string.
+
+    Optionally resize the image before encoding to improve performance for downstream processing.
+
+    :param bytestream: ByteStream containing the image data.
+    :param size: If provided, resizes the image to fit within the specified dimensions (width, height) while
+        maintaining aspect ratio. This reduces file size, memory usage, and processing time, which is beneficial
+        when working with models that have resolution constraints or when transmitting images to remote services.
+
+    :returns:
+        A tuple (mime_type, base64_str), where:
+        - mime_type (str): The MIME type of the encoded image, determined from the original data or image content.
+          Defaults to 'image/jpeg' if the type cannot be reliably identified.
+        - base64_str (str): The base64-encoded string representation of the (optionally resized) image.
+    """
+    if size is None:
```

Comment:
there might be an inconsistency related to`mime_type` here:
- if `size` is None, we return `bytestream.mime_type` which can be `None` if the original source was a `ByteStream` (not a path)
- otherwise, we try to infer it and return `"image/jpeg"` if we fail to infer it (this is also done because we need to specify a `form` in `image.save`)

(I'm not sure of my reasoning because I see that mypy is not raising any error.)

For consistency, I would prefer to return `"image/jpeg"` also when `bytestream.mime_type` is `None`. Makes sense?

Reply:
Ok, I understand and seem to make sense.

What I would do:
- change the return type of `encode_image_to_base64` to `Tuple[Optional[str], str]`
- try to explain this aspect in the docstring
- (maybe) `if bytestream.mime_type is None`, show a warning similar to the `if not inferred_mime_type` below

Reply:
Done! I've also added a warning message. I think it could be improved but I'm not entirely sure how. Mostly surrounding how to tell the user how to provide the missing mime type.

Reply:
thanks!